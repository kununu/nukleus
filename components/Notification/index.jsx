import React, {Component, PropTypes} from 'react';

import styles from './index.scss';

export default class Notification extends Component {
  static propTypes = {
    closeIcon: PropTypes.element,
    closeMethod: PropTypes.oneOf(['onClick', 'onTimeout']),
    duration: PropTypes.number,
    icon: PropTypes.element,
    message: PropTypes.string.isRequired,
    onCloseClick: PropTypes.func,
    type: PropTypes.oneOf([
      'error',
      'success'
    ]),
    visible: PropTypes.bool
  };

  static defaultProps = {
    closeIcon: (
      <svg
        version="1.1"
        xmlns="http://www.w3.org/2000/svg"
        xmlnsXlink="http://www.w3.org/1999/xlink" x="0px" y="0px"
        viewBox="0 0 68.3 61.2"
        xmlSpace="preserve">
        <path
          d="M55,43.3c0,0.9-0.3,1.7-1,2.4l-4.8,4.8c-0.7,0.7-1.5,1-2.4,1c-0.9,0-1.7-0.3-2.4-1L34.1,40.2L23.8,50.5
            c-0.7,0.7-1.5,1-2.4,1c-0.9,0-1.7-0.3-2.4-1l-4.8-4.8c-0.7-0.7-1-1.5-1-2.4s0.3-1.7,1-2.4l10.3-10.3L14.2,20.3
            c-0.7-0.7-1-1.5-1-2.4s0.3-1.7,1-2.4l4.8-4.8c0.7-0.7,1.5-1,2.4-1c0.9,0,1.7,0.3,2.4,1L34.1,21l10.3-10.3c0.7-0.7,1.5-1,2.4-1
            c0.9,0,1.7,0.3,2.4,1l4.8,4.8c0.7,0.7,1,1.5,1,2.4s-0.3,1.7-1,2.4L43.7,30.6L54,40.9C54.7,41.6,55,42.4,55,43.3z" />
      </svg>
    ),
    closeMethod: 'onClick',
    duration: 5000,
    icon: null,
    message: '',
    onCloseClick: null,
    type: 'success',
    visible: false
  }

  state = {
    visible: this.props.visible
  };

  componentWillReceiveProps (nextProps) {
    if (this.newNotificationWillOverrideExisting(
      nextProps.message
    )) {
      // Briefly set visible state to false to show animation again
      this.setState({visible: false});
      setTimeout(() => this.setState({visible: true}), 100);
    } else {
      this.setState({visible: nextProps.visible});
    }
  }

  onClickClose = () => {
    this.hideNotification();
    if (this.props.onCloseClick) {
      this.props.onCloseClick();
    }
  };

  onTimeout () {
    if (this.props.closeMethod === 'onTimeout') this.hideNotification();
  }

  newNotificationWillOverrideExisting (message) {
    return (
      this.props.visible &&
      this.props.message !== message
    );
  }

  hideNotification () {
    this.setState({visible: false});
  }

  isError () {
    return this.props.type === 'error';
  }

  render () {
    const {
      closeIcon,
      duration,
      icon,
      message,
      type,
      visible
    } = this.props;

    if (visible && !this.isError()) {
      setTimeout(() => {
        this.onTimeout();
      }, duration);
    }

    return (
      <div
        className={`
          ${styles.container}
          ${styles[this.state.visible ? 'visible' : 'hidden']}
          ${this.isError() && styles.error}
        `}>

        <p>
          {this.props.closeMethod === 'onClick' &&
            <button
              className={styles.closeButton}
              onClick={this.onClickClose}>
              {closeIcon}
            </button> }

          {icon &&
            <span className={styles[type]}>
              {icon}
            </span>
          }

          {message}
        </p>
      </div>
    );
  }
}
