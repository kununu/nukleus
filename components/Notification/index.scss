@import '~kununu-theme/scss/base/variables';

.container {
  width: 320px;
  padding: 20px;
  height: auto;
  min-height: 80px;
  background-color: $teal;
  border-radius: $border-radius-large;
  border: 1px solid $teal-dark;
  position: fixed;
  left: 20px;
  z-index: $zindex-popover;
  display: none;

  &,
  * {
    box-sizing: border-box;
  }

  span {
    padding-right: 8px;
  }

  i {
    font-size: 18px;
  }

  p {
    color: $gray-base-30;
    margin-bottom: 6px;
    letter-spacing: -.035em;
    margin-top: $baseline-grid;
  }
}

.closeButton {
  color: $gray-base-30;
  background-color: transparent;
  border: 0;
  position: absolute;
  right: 10px;
  top: 10px;
  cursor: pointer;
  outline: 0;
}

.container.error {
  background-color: $brand-danger;
  border-color: $red-dark;

  p {
    color: $white;
    font-weight: $font-weight-semibold;
  }

  .closeButton {
    color: $red-dark;
  }
}

.success {
  color: $green;
}

.container.visible {
  animation-name: slideUp;
  animation-duration: 1s;
  animation-fill-mode: both;
  opacity: 1;
  bottom: 20px;
  display: block;
}

.container.hidden {
  display: none;
}

@keyframes slideUp {
  0%,
  60%,
  75%,
  90%,
  100% {
    transition-timing-function: cubic-bezier(.215, .61, .355, 1);
  }

  0% {
    opacity: 0;
    transform: translateY(3000px);
  }

  60% {
    transform: translateY(-20px);
  }

  75% {
    transform: translateY(10px);
  }

  90% {
    transform: translateY(-5px);
  }

  100% {
    transform: translateY(0);
  }
}

