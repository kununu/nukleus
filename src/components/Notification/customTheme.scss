@import '~@kununu/kununu-theme-v2/scss/base/variables';

.notificationContainer {
  width: 320px;
  padding: 20px;
  height: auto;
  min-height: 80px;
  background-color: $green;
  border-radius: 10px;
  position: fixed;
  left: 20px;
  z-index: 1060;
  display: none;

  &,
  * {
    box-sizing: border-box;
  }

  span {
    padding-right: 8px;
  }

  i {
    font-size: 18px;
  }

  p {
    color: $white;
    margin-bottom: 6px;
    margin-top: $baseline-grid;
  }
}

.notificationCloseButton {
  background-color: transparent;
  border: 0;
  position: absolute;
  right: 5px;
  top: 10px;
  cursor: pointer;
  outline: 0;

  svg {
    fill: $white;
    width: 16px;
    height: 16px;
  }
}

.notificationContainer.visible {
  animation-name: slideRight;
  animation-duration: 1s;
  animation-fill-mode: both;
  opacity: 1;
  bottom: 20px;
  display: block;
}

.notificationContainer.hidden {
  display: none;
}

@keyframes slideRight {
  0%,
  60%,
  75%,
  90%,
  100% {
    transition-timing-function: cubic-bezier(.215, .61, .355, 1);
  }

  0% {
    opacity: 0;
    transform: translateX(3000px);
  }

  60% {
    transform: translateX(-20px);
  }

  75% {
    transform: translateX(10px);
  }

  90% {
    transform: translateX(-5px);
  }

  100% {
    transform: translateX(0);
  }
}
