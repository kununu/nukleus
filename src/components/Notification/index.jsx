import React from 'react';
import PropTypes from 'prop-types';

import ThemeContext from 'utils/themeContext';
import themeable from 'utils/theming';

import styles from './index.module.scss';

export default class Notification extends React.Component {
  static propTypes = {
    closeIcon: PropTypes.element,
    closeMethod: PropTypes.oneOf(['onClick', 'onTimeout']),
    duration: PropTypes.number,
    icon: PropTypes.element,
    message: PropTypes.string.isRequired,
    onClose: PropTypes.func,
    onCloseClick (props, propName) { // eslint-disable-line
      if (props[propName]) {
        return new Error('This prop is marked as deprecated and will be removed on upcoming release. ' +
          'Please use `onClose` instead.');
      }
    },
    type: PropTypes.oneOf([
      'error',
      'success',
    ]),
    visible: PropTypes.bool,
  };

  static defaultProps = {
    closeIcon: (
      <svg
        version="1.1"
        xmlns="http://www.w3.org/2000/svg"
        xmlnsXlink="http://www.w3.org/1999/xlink"
        x="0px"
        y="0px"
        viewBox="0 0 68.3 61.2"
        xmlSpace="preserve"
      >
        <path
          d="M55,43.3c0,0.9-0.3,1.7-1,2.4l-4.8,4.8c-0.7,0.7-1.5,1-2.4,1c-0.9,0-1.7-0.3-2.4-1L34.1,40.2L23.8,50.5
            c-0.7,0.7-1.5,1-2.4,1c-0.9,0-1.7-0.3-2.4-1l-4.8-4.8c-0.7-0.7-1-1.5-1-2.4s0.3-1.7,1-2.4l10.3-10.3L14.2,20.3
            c-0.7-0.7-1-1.5-1-2.4s0.3-1.7,1-2.4l4.8-4.8c0.7-0.7,1.5-1,2.4-1c0.9,0,1.7,0.3,2.4,1L34.1,21l10.3-10.3c0.7-0.7,1.5-1,2.4-1
            c0.9,0,1.7,0.3,2.4,1l4.8,4.8c0.7,0.7,1,1.5,1,2.4s-0.3,1.7-1,2.4L43.7,30.6L54,40.9C54.7,41.6,55,42.4,55,43.3z"
        />
      </svg>
    ),
    closeMethod: 'onClick',
    duration: 5000,
    icon: null,
    onClose: () => {},
    type: 'success',
    visible: false,
  };

  state = {visible: this.props.visible}; // eslint-disable-line react/destructuring-assignment

  UNSAFE_componentWillReceiveProps (nextProps) { // eslint-disable-line
    if (this.newNotificationWillOverrideExisting(nextProps.message)) {
      // Briefly set visible state to false to show animation again
      this.setState({visible: false});
      setTimeout(() => this.setState({visible: true}), 100);
    } else {
      this.setState({visible: nextProps.visible});
    }
  }

  componentWillUnmount () {
    if (this.timeout) {
      clearTimeout(this.timeout);
    }
  }

  onClickClose = () => {
    const {onClose} = this.props;

    this.hideNotification();
    onClose();
  };

  onTimeout () {
    const {
      closeMethod,
      onClose,
    } = this.props;

    if (closeMethod === 'onTimeout') {
      this.hideNotification();
      onClose();
    }
  }

  newNotificationWillOverrideExisting (message) {
    const {
      message: pmessage,
      visible,
    } = this.props;

    return (
      visible &&
      pmessage !== message
    );
  }

  hideNotification () {
    this.setState({visible: false});
  }

  isError () {
    const {type} = this.props;

    return type === 'error';
  }

  render () {
    const {
      closeIcon,
      closeMethod,
      duration,
      icon,
      message,
    } = this.props;

    const {visible} = this.state;

    if (visible) {
      this.timeout = setTimeout(() => {
        this.onTimeout();
      }, duration);
    }

    return (
      <ThemeContext.Consumer>
        {(context) => {
          const theme = themeable({...styles, ...context});

          return (
            <div
              className={theme('notificationContainer', `${visible ? 'visible' : 'hidden'}`, `${this.isError() ? 'notificationError' : ''}`)}
            >

              <p>
                {closeMethod === 'onClick' && (
                <button
                  className={theme('notificationCloseButton')}
                  onClick={this.onClickClose}
                  type="button"
                >
                  {closeIcon}
                </button>
                )}

                {icon && (
                <span className={theme('notificationIcon')}>
                  {icon}
                </span>
                )}

                {message}
              </p>
            </div>
          );
        }}
      </ThemeContext.Consumer>
    );
  }
}
