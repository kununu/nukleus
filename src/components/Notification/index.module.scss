@import '~@kununu/kununu-theme-v2/scss/base/variables';

.notificationContainer {
  width: 320px;
  padding: 20px;
  height: auto;
  min-height: 80px;
  background-color: $teal;
  border-radius: $border-radius-large;
  border: 1px solid $teal-dark;
  position: fixed;
  left: 20px;
  z-index: 1060;
  display: none;

  &,
  * {
    box-sizing: border-box;
  }

  span {
    padding-right: 8px;
  }

  i {
    font-size: 18px;
  }

  p {
    color: $gray-base-30;
    margin-bottom: 6px;
    margin-top: $baseline-grid;
  }

  .notificationIcon {
    color: $green;
  }
}

.notificationCloseButton {
  background-color: transparent;
  border: 0;
  position: absolute;
  right: 5px;
  top: 10px;
  cursor: pointer;
  outline: 0;

  svg {
    fill: $gray-base-30;
    width: 16px;
    height: 16px;
  }
}

.notificationContainer.notificationError {
  background-color: $brand-danger;
  border-color: $red-dark;

  p {
    color: $white;
    font-weight: $font-weight-semibold;
  }

  .notificationCloseButton {
    color: $red-dark;

    svg {
      fill: $red-dark;
    }
  }

  .notificationIcon {
    color: $white;
  }
}

.notificationContainer.visible {
  animation-name: slideUp;
  animation-duration: 1s;
  animation-fill-mode: both;
  opacity: 1;
  bottom: 20px;
  display: block;
}

.notificationContainer.hidden {
  display: none;
}

@keyframes slideUp {
  0%,
  60%,
  75%,
  90%,
  100% {
    transition-timing-function: cubic-bezier(.215, .61, .355, 1);
  }

  0% {
    opacity: 0;
    transform: translateY(3000px);
  }

  60% {
    transform: translateY(-20px);
  }

  75% {
    transform: translateY(10px);
  }

  90% {
    transform: translateY(-5px);
  }

  100% {
    transform: translateY(0);
  }
}
