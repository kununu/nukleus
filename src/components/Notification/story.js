import React from 'react';
import {storiesOf} from '@storybook/react';
import {defaultOptions} from 'defaultOptions'; // eslint-disable-line import/no-unresolved
import {withOptions} from '@storybook/addon-options';
import {withInfo} from '@storybook/addon-info';
import {action} from '@storybook/addon-actions';
import {withKnobs, text, number} from '@storybook/addon-knobs/react';

import ThemeProvider from '../ThemeProvider';

import customTheme from './customTheme.scss';

import Notification from './index';

storiesOf('Notification', module)
  .addDecorator(withOptions({...defaultOptions}))
  .addDecorator(withKnobs)
  .addDecorator((story, context) => withInfo('The notification component!')(story)(context))
  .add('error', () => (
    <Notification
      type="error"
      closeMethod={text('closeMethod', 'onTimeout')}
      onClose={action('onClose')}
      message={text('message', 'Hi, I am an error notification box')}
      icon={<i className="fa fa-cross" />}
      visible
    />
  ))
  .add('success', () => (
    <Notification
      onClose={action('onClose')}
      message={text('message', 'Hi, I am a success notification box')}
      icon={<i className="fa fa-check" />}
      visible
    />
  ))
  .add('time out', () => (
    <Notification
      closeMethod="onTimeout"
      duraction={number('duration', 3000)}
      onClose={action('onClose')}
      message={text('message', 'Hi, I am a success notification box')}
      icon={<i className="fa fa-check" />}
      visible
    />
  ))
  .add('custom', () => (
    <ThemeProvider theme={customTheme}>
      <Notification
        onClose={action('onClose')}
        message={text('message', 'Hi, I am a success notification box')}
        icon={<i className="fa fa-check" />}
        visible
      />
    </ThemeProvider>
  ));
