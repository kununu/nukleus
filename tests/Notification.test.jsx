import React from 'react';
import renderer from 'react-test-renderer';
import {mount, shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import Notification from 'Notification'; // eslint-disable-line import/no-unresolved, import/extensions, import/no-extraneous-dependencies

const notification = (
  <Notification
    message="Test"
    visible />
);
const timeoutNotification = (
  <Notification
    message="Test"
    closeMethod="onTimeout"
    duration={8000}
    visible />
);
const successNotification = (
  <Notification
    icon={<i className="fa fa-check" />}
    message="Test"
    visible />
);

test('Renders Notification without crashing', () => {
  const component = renderer.create(notification);
  expect(component.toJSON()).toMatchSnapshot();
});

test('Closes when the close button is clicked', () => {
  const component = mount(notification);
  component.find('.closeButton').simulate('click');
  expect(toJson(component)).toMatchSnapshot();
});

test('Does not close on click when the close method is timeout', () => {
  const component = mount(timeoutNotification);
  expect(component.find('.closeButton').props.onClick).toThrow();
  expect(toJson(component)).toMatchSnapshot();
});

test('It closes on timeout when the close method is timeout', () => {
  const component = shallow(timeoutNotification);
  expect(component.state('visible')).toEqual(true);
  component.instance().onTimeout();
  expect(component.state('visible')).toEqual(false);
});

test('It shows content icon when defined in props', () => {
  const component = renderer.create(successNotification);
  expect(component.toJSON()).toMatchSnapshot();
});
