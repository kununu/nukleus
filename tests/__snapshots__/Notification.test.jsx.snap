exports[`test Closes when the close button is clicked 1`] = `
<Notification
  closeMethod="onClick"
  duration={5000}
  message="Test"
  type="SUCCESS"
  visible={true}>
  <div
    className="
            container
            hidden
            false
          ">
    <p>
      <button
        className="fa fa-times closeButton"
        onClick={[Function]} />
      <span>
        <i
          className="fa fa-check success" />
      </span>
      Test
    </p>
  </div>
</Notification>
`;

exports[`test Does not close on click when the close method is timeout 1`] = `
<Notification
  closeMethod="onTimeout"
  duration={8000}
  message="Test"
  type="SUCCESS"
  visible={true}>
  <div
    className="
            container
            visible
            false
          ">
    <p>
      <span>
        <i
          className="fa fa-check success" />
      </span>
      Test
    </p>
  </div>
</Notification>
`;

exports[`test Renders Notification without crashing 1`] = `
<div
  className="
          container
          visible
          false
        ">
  <p>
    <button
      className="fa fa-times closeButton"
      onClick={[Function]} />
    <span>
      <i
        className="fa fa-check success" />
    </span>
    Test
  </p>
</div>
`;
